File: Two Pointers/3. 3Sum/solution.rb
def three_sum(nums)
    nums.sort!
    result = []
    nums.each_with_index do |n, i|
        next if i > 0 && n == nums[i - 1]
        l, r = i + 1, nums.length - 1
        while l < r
            three_sum = n + nums[l] + nums[r]
            if three_sum > 0
                r -= 1
            elsif three_sum < 0
                l += 1
            else
                result.push([n, nums[l], nums[r]])
                l += 1
                while nums[l] == nums[l - 1] && l < r
                    l += 1
                end
            end
        end
    end

    return result
end

File: Two Pointers/3. 3Sum/solution.scala
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        triplets = set()
        for i in range(len(nums) - 2):
            if i > 0 and nums[i] == nums[i - 1]:
                continue  # Skip duplicate values of the first number
            firstNum = nums[i]
            j, k = i + 1, len(nums) - 1
            while j < k:
                secondNum, thirdNum = nums[j], nums[k]
                potentialSum = firstNum + secondNum + thirdNum
                if potentialSum > 0:
                    k -= 1
                elif potentialSum < 0:
                    j += 1
                else:
                    triplets.add((firstNum, secondNum, thirdNum))
                    j, k = j + 1, k - 1
                    while j < k and nums[j] == nums[j - 1]:
                        j += 1  # Skip duplicate values of the second number
                    while j < k and nums[k] == nums[k + 1]:
                        k -= 1  # Skip duplicate values of the third number
        return triplets
